tc/core/cuda/cuda_rtc.cc

```diff
@@ -60,30 +64,87 @@ void checkOrCreateContext() {
   }
 }
 
-std::unique_ptr<CudaRTCFunction> CudaRTCFunction::Compile(
-    const std::string& name,
-    const std::string& source) {
-  std::unique_ptr<CudaRTCFunction> res(new CudaRTCFunction());
-  res->specializedName = name;
-  res->cleared_ = false;
-
-  if (FLAGS_debug_tc_mapper) {
-    LOG(INFO) << "NVRTC function source:\n" << source;
-  }
-  // Actually do the compiling.
-  nvrtcProgram prog;
-  TC_NVRTC_CHECK(
-      nvrtcCreateProgram(&prog, source.c_str(), nullptr, 0, nullptr, nullptr));
-
-  // Get the architecture of the current device.
-  int device, minor, major;
+namespace {
+static std::tuple<int, int, int> getCudaArchitecture() {
+  int device, major, minor;
   CUdevice deviceHandle;
   TC_CUDA_RUNTIMEAPI_ENFORCE(cudaGetDevice(&device));
   TC_CUDA_DRIVERAPI_ENFORCE(cuDeviceGet(&deviceHandle, device));
   TC_CUDA_DRIVERAPI_ENFORCE(cuDeviceGetAttribute(
       &major, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR, deviceHandle));
   TC_CUDA_DRIVERAPI_ENFORCE(cuDeviceGetAttribute(
       &minor, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MINOR, deviceHandle));
+  return std::tuple<int, int, int>(device, major, minor);
+}
+
+static std::string llvmCompile(
+    const std::string& name,
+    const std::string& source) {
+  int device, major, minor;
+  std::tie(device, major, minor) = getCudaArchitecture();
+
+  std::string pat("/tmp/cudaXXXXXX");
+  std::vector<char> ifn(pat.begin(), pat.end());
+  TC_CHECK_GE(mkstemp(ifn.data()), 0); // string.c_str is const char*
+  std::string inputFileName(ifn.begin(), ifn.end());
+  tc::ScopeGuard sgi([&]() { std::remove(inputFileName.c_str()); });
+  {
+    std::ofstream ostream(inputFileName, std::ios::binary);
+    ostream << source;
+  }
+
+  std::string arch = "sm_" + std::to_string(major) + std::to_string(minor);
+  std::string outputClangFile = inputFileName + "-clang.ll";
+  std::string outputLinkFile = inputFileName + "-link.ll";
+  std::string outputOptFile = inputFileName + "-opt.ll";
+  std::string outputPtxFile = inputFileName + ".s";
+  tc::ScopeGuard sgo([&]() {
+    std::remove(outputClangFile.c_str());
+    std::remove(outputLinkFile.c_str());
+    std::remove(outputOptFile.c_str());
+    std::remove(outputPtxFile.c_str());
+  });
+
+  std::string cmdLlvmIr = std::string(TC_STRINGIFY(TC_LLVM_BIN_DIR)) +
```

Comment:
I'd use stringstreams here

Reply:
or even have a "checkedSystemCall" variadic function that takes an arbitrary number of strings it concatenates internally, calls `std::system` and checks if it was ok